File: oe/5w/priorityqueue.cxx
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

/**/
#include "edx-io.hpp"
#define cout io
#define cin io
/**/


void siftdown(vector<long long> &heap, long long startpos, long long pos) {
    long long newitem = heap[pos];
    while (pos > startpos) {
        long long parentpos = (pos - 1) / 2;
        long long parent = heap[parentpos];
        if (parent > newitem) {
            heap[pos] = parent;
            pos = parentpos;
            continue;
        }
        break;
    }
    heap[pos] = newitem;
}

void siftup(vector<long long> &heap, long long pos) {
    long long endpos = heap.size();
    long long startpos = pos;
    long long newitem = heap[pos];
    long long childpos = 2*pos + 1;
    while (childpos < endpos) {
        long long rightpos = childpos + 1;
        if ((rightpos < endpos) && !( heap[childpos] < heap[rightpos])) { //TODO
            childpos = rightpos;
            // cout << "A";
        }
        heap[pos] = heap[childpos];
        pos = childpos;
        childpos = 2*pos + 1;
    }
    heap[pos] = newitem;
    siftdown(heap, startpos, pos);
}

void heappush(vector<long long> &heap, long long item){
    heap.push_back(item);
    siftdown(heap, 0, heap.size()-1);
}

long long heappop(vector<long long> &heap){
    long long lastelt = heap.back();
    heap.pop_back();
    if (!heap.empty()) {
        long long returnitem = heap[0];
        heap[0] = lastelt;
        siftup(heap, 0);
        return returnitem;
    }
    return lastelt;
}

int main() {
    long long N;
    cin >> N;

    vector<long long> heap;
    heap.reserve(1000000);
    long long* array = new long long[N+1];
    char action;
    long long a;

    for (long long i = 0; i < N; i++) {
        cin >> action;
        switch (action)
        {
            case 'A':
                cin >> a;
                heappush(heap, a);
                array[i] = a;
                break;
            case 'X':
                if (heap.size() == 0) {
                    cout << '*' << '\n';
                }
                else {
                    long long pop = heappop(heap);
                    cout << pop << '\n';
                }
                break;
            case 'D':
                long long newNum;
                cin >> a;
                cin >> newNum;
                long long oldNum = array[a-1];
                array[a-1] = newNum;
                long long ind = find(heap.begin(), heap.end(), oldNum) - heap.begin();
                heap[ind] = newNum;
                siftdown(heap, 0, ind);
                // siftup(heap, 0);
                break;
        }
    }
    return 0;
}

File: oe/5w/priorityqueue.py
from edx_io import edx_io

def heappush(heap, item):
    heap.append(item)
    siftdown(heap, 0, len(heap)-1)

def heappop(heap):
    """Pop the smallest item off the heap, maintaining the heap invariant."""
    lastelt = heap.pop()    # raises appropriate IndexError if heap is empty
    if heap:
        returnitem = heap[0]
        heap[0] = lastelt
        siftup(heap, 0)
        return returnitem
    return lastelt

def siftup(heap, pos):
    endpos = len(heap)
    startpos = pos
    newitem = heap[pos]
    # Bubble up the smaller child until hitting a leaf.
    childpos = 2*pos + 1    # leftmost child position
    while childpos < endpos:
        # Set childpos to index of smaller child.
        rightpos = childpos + 1
        if rightpos < endpos and not heap[childpos] < heap[rightpos]:
            childpos = rightpos
            print("A")
        # Move the smaller child up.
        heap[pos] = heap[childpos]
        pos = childpos
        childpos = 2*pos + 1
    # The leaf at pos is empty now.  Put newitem there, and bubble it up
    # to its final resting place (by sifting its parents down).
    heap[pos] = newitem
    siftdown(heap, startpos, pos)

def siftdown(heap, startpos, pos):
    newitem = heap[pos]
    # Follow the path to the root, moving parents down until finding a place
    # newitem fits.
    while pos > startpos:
        parentpos = (pos - 1) >> 1
        parent = heap[parentpos]
        if newitem < parent:
            heap[pos] = parent
            pos = parentpos
            continue
        break
    heap[pos] = newitem

def main(io):
	heap = []
	# fo= open("output.txt","w")
	# fi = open('input.txt', 'r')
	n = io.next_int()
	arr = [0]*(10**6)
	for i in range(0, n):
		inp = io.next_token()
		if inp == b'A':
			p = io.next_int()
			heappush(heap, p)
			arr[i] = p
		elif inp == b'X':
			if len(heap)==0:
				io.writeln("*")
				continue
			pop = heappop(heap)
			io.writeln(str(pop))
			# fo.write(str(pop))
			# fo.write("\n")
			# TODO
		elif inp == b'D':
			p = io.next_int()
			oldNum = arr[p-1]
			newNum = io.next_int()
			arr[p-1] = newNum
			ind = heap.index(oldNum)
			heap[ind] = newNum
			# heapq._siftdown(heap, 0, ind)
			# heapq._siftup(heap, ind)
			siftdown(heap, 0, ind)
		else:
			print(inp)

if __name__ == "__main__":
	with edx_io() as io:
	    main(io)